velox/expression/ComplexProxyTypes.h

```diff
@@ -228,4 +233,166 @@ class ArrayProxy {
   template <typename T>
   friend class SimpleFunctionAdapter;
 };
+
+// The object passed to the simple function interface that represent a single
+// output map entry.
+//
+// General Interface:
+// - add_item()  : return references to key and value writers as tuple.
+// - add_null()  : return key writer, set value to null.
+// - size()      : return the size of the map.
+// The interface does not guarantee that duplicates not written.
+template <typename K, typename V>
+class MapWriter {
+  using key_writer_t = VectorWriter<K, void>;
+  using value_writer_t = VectorWriter<V, void>;
+
+  using key_element_t = typename key_writer_t::exec_out_t;
+  using value_element_t = typename value_writer_t::exec_out_t;
+
+  // Provide std::interface only if both key and value are eligible types.
+  static constexpr bool std_interface =
+      provide_std_interface<K> && provide_std_interface<V>;
+
+ public:
+  // `size` is with respect to the current size of the map being written.
+  void reserve(vector_size_t size) {
+    auto currentSize = size + innerOffset_;
+
+    if (UNLIKELY(currentSize > capacity_)) {
+      while (currentSize > capacity_) {
+        capacity_ = capacity_ * 2;
+      }
+      keysWriter_->ensureSize(capacity_);
+      valuesWriter_->ensureSize(capacity_);
+    }
+  }
+
+  // Add a new not null item to the map, increasing its size by 1.
+  std::tuple<key_element_t&, value_element_t&> add_item() {
+    resize(length_ + 1);
+    return std::tie(lastKeyWriter(), lastValueWriter());
+  }
+
+  // Add a new null value, return the key writer.
+  key_element_t& add_null() {
+    resize(length_ + 1);
+    valuesVector_->setNull(indexOfLast(), true);
+    return lastKeyWriter();
+  }
+
+  vector_size_t size() {
+    return length_;
+  }
+
+ private:
+  // Make sure user do not use those.
+  MapWriter<K, V>() = default;
+
+  MapWriter<K, V>(const MapWriter<K, V>&) = default;
+
+  MapWriter<K, V>& operator=(const MapWriter<K, V>&) = default;
+
+  vector_size_t indexOfLast() {
+    return innerOffset_ + length_ - 1;
+  }
+
+  // Should be called by the user (VectorWriter) when writing is done to commit
+  // last item if needed.
+  void finalize() {
+    commitMostRecentChildItem();
+    innerOffset_ += length_;
+    length_ = 0;
+  }
+
+  // 'size' is with respect to the current size of the array being written.
+  void resize(vector_size_t size) {
+    commitMostRecentChildItem();
+    reserve(size);
+    length_ = size;
+  }
+
+  void commitMostRecentChildItem() {
+    if constexpr (requires_commit<K>) {
```

Comment:
Why do you distinguish between writers that require commit and those that don't?

There's a lot of duplicate logic from VectorWriter here for types that don't need commit, even though they support commit and could use the same logic, and it seems like it breaks abstractions VectorWriter provides (and hence is a little risky).

Is the performance of this function the key benefit?  Admittedly it's a no-op if we don't require commit.  I'm just double checking you confirmed it's significant.

Reply:
basically i do not require commit for primitives and bypass the vector writer protocol  speedup

simple_emplace                                               1.75ms   570.59
simple_general                                               1.76ms   568.41

simple_emplace                                               2.21ms   451.92
simple_general                                               2.50ms   399.69